### PR TITLE
fix: configure proper pages directory for next application

### LINKED ISSUE #17898: Component testing + Next JS using @cypress/react/plugins/next configures the 'pages' directory wrongly
### Current behavior

I believe that the internal configuration from the `@cypress/react/plugins/next` configures 'nextjs' to think that the pages directory is the 'root folder of the project' rather than `${rootFolder}/pages`.

If you look at this line:
https://github.com/cypress-io/cypress/blob/develop/npm/react/plugins/next/findNextWebpackConfig.js#L29
Screen Shot 2021-08-26 at 4 06 35 PM

The findNextWebpackConfig util is setting the pagesDir to the root of the project. 

#### Consequences
Because of that I'm seeing that next js webpack config is treating every folder in my root as a 'pages' folder, trying to bundle them as if they were pages. I bumped into this because I started seeing the following error while running component tests:
```
Syntax error: Using `export * from '...'` in a page is disallowed. Please use `export { default } from '...'` instead.
Read more: https://nextjs.org/docs/messages/export-all-in-page
```
At first I thought I had some 'export * from ...'  inside a 'page' but that was not the case. I found that a file inside `${root}/src/UILibrary/icons` was doing `export * from ...`. And because of how the cypress next plugin is configuring webpack my `/src/UILibrary/icons` is being treated as a page by next+webpack.

### Desired behavior

Not fully certain of the solution. Probably allow to pass that configuration instead of assuming that `pagesDir` is `config.projectRoot`. So we can do the following inside our plugins config file:
```
const injectDevServer = require("@cypress/react/plugins/next");

module.exports = (on, config) => {
  if (config.testingType === "component") {
    injectDevServer(on, {...config, pagesDir: 'path to my pages dir' });
  }

  return config;
};
```

Optionally we could have a better default pointing to `${root}/pages` (if it exists) or to `${root}/src/pages` (which is the second option that nextjs recognizes as convention. 

### Test code to reproduce

I forked the webpack example from cypress docs and added a `export * from ...` to trigger the error. (this happens in webpack 4 or webpack 5 config the same. I used the webpack 4 because I wanted to use the official example published by cypress with the minimum amount of changes).

Here the fork and branch where to try it: https://github.com/rafaelchiti/cypress-component-examples/tree/bug_repro

To see the error make sure you are on that branch `bug_repo` and run:
`cd nextjs-webpack-4` to focus on the next example
`yarn install --pure-lockfile`
`DEBUG=* yarn run cypress run-ct`
If you don't set a wide debug filter option then the error gets swollen (probably there is a more limited debug config that still prints the error). Then after running that you should find in the terminal an error like:
```
Syntax error: Using `export * from '...'` in a page is disallowed. Please use `export { default } from '...'` instead.
Read more: https://nextjs.org/docs/messages/export-all-in-page
```

But if you look in the code there is no `export * from ...` inside the pages directory. But there is though an `export * from ...` inside the `src/icons` directory (which I added to test the problem). 

### Cypress Version

7.2.0 also tried on 8.3.0

### Other

_No response_

### PR DESCRIPTION
**Pull request by ZachJW34**
- Closes #17898

### User facing changelog
NA

### Additional details
Our Next.js plugin was not pointing to the correct pagesDir, causing errors to be thrown if files under projectRoot did not follow the pages export convention. This PR sets the `pagesDir` by looking for either `{root}/pages` or `{root}/src/pages` (Next.js conventions) and falling back to `{root}` if not found. 

I tested this by checking out https://github.com/cypress-io/cypress-component-testing-examples and:

1. `cd ./create-next-app`
2. `yarn`
3.  `mkdir -p ./src/components`
3. `echo "export const HelloWorld = () => 'Hello World!'" >> ./src/components/HelloWorld.js`
4. `echo "export * from './HelloWorld'" >> ./src/components/index.js`
5. Add `import { HelloWorld } from '../src/components'` to `./pages/index.js`
6. `npx cypress run-ct` and see that it fails (You can run DEBUG=* npx cypress run-ct to see logs)
7. Copy new `plugins` folder from this branch (or entire built package) into `./node_modules/@cypress/react`
8. `npx cypress run-ct` and see that it passes 

### How has the user experience changed?
NA

### PR Tasks

- [x] Have tests been added/updated?
- [ ] Has the original issue or this PR been tagged with a release in ZenHub? 
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- [ ] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

File: npm/react/package.json [hunk 1]
```diff
@@ -12,6 +12,9 @@
     "cy:run:debug": "node --inspect-brk ../../scripts/start.js --component-testing --run-project ${PWD}",
     "pretest": "yarn transpile",
     "test": "yarn cy:run",
+    "yarn": "yarn test-unit && yarn test-component",
+    "test-unit": "mocha test/**/*.spec.js",
```
Comment (JessicaSachs) [inline npm/react/package.json:16]:
```suggestion
    "test-unit": "mocha test/**/*.spec.*",
```

File: npm/react/plugins/next/findPagesDir.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+  pagesDir = path.join(projectRoot, 'src/pages')
+  if (existsSync(pagesDir)) {
+    return pagesDir
+  }
+
+  return projectRoot
+}
+
+module.exports.findPagesDir = findPagesDir
```
Comment (JessicaSachs) [inline npm/react/plugins/next/findPagesDir.js:38]: For example, in TS this would be 

```ts
export function findPagesDir() { /* ... */ }
```

**Review by JessicaSachs — CHANGES_REQUESTED** [review]
This looks good, but we need to make sure that our code is written ts-first. If you have a sec to convert the new Javascript files into Typescript, that would be super.

File: npm/react/package.json [hunk 3]
```diff
@@ -12,6 +12,9 @@
     "cy:run:debug": "node --inspect-brk ../../scripts/start.js --component-testing --run-project ${PWD}",
     "pretest": "yarn transpile",
     "test": "yarn cy:run",
+    "yarn": "yarn test-unit && yarn test-component",
```
Comment (lmiller1990) [inline npm/react/package.json:15]: Is this step supposed to be called `yarn`? This doesn't seem correct.

**Review by lmiller1990 — COMMENTED** [review]
Nice job figuring this out - just need some TS and to clarify what the `yarn` command is for (`yarn`  does not seem like a great name for a script in package.json).

File: npm/react/package.json (hunk 3 above)
Comment (ZachJW34) [reply npm/react/package.json:15]: Good catch. I meant to replace line 14 with `"test": "yarn test-unit && yarn test-component",`

**Comment by ZachJW34** [comment]
@lmiller1990 @JessicaSachs Is the `appveyor` CI failure something I should be concerned about? I can't reproduce the error locally. If I do need to make a change to how I'm compiling the typescript I see two ways:

1. Move the `findPagesDir.ts` into the src/ directory. This makes sense because the `rootDir` is set as `src` so that's where TS files should be located.
2. Create a separate tsconfig inside of plugins.

There are more drastic changes we could make but I didn't want to change too much just for a small bug fix. Thoughts?

**Review by JessicaSachs — DISMISSED** [review]
LGTM. No, Appveyor sometimes doesn't retrigger. It's usually fine, though.

**Comment by lmiller1990** [comment]
Seems the build is broken. You should be able to reproduce locally with `yarn build` @ZachJW34

**Comment by ZachJW34** [comment]
@lmiller1990 Looks like windows didn't like the glob pattern for tsc cli. I added a tsconfig for plugins so it should be good now.

**Review by lmiller1990 — APPROVED** [review]

**Pull request merged**